File: src/main/java/eu/fabiool/cnpj/file/parsing/lib/model/Socio.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package eu.fabiool.cnpj.file.parsing.lib.model;



import static java.util.Arrays.copyOfRange;
/**
 *
 * @author Fabio
 */
public class Socio extends AbstractInfoCnpj {

    private final char tipoDoRegistro;
    private final char indicadorFullDiario;
    private final char tipoAtualização;
    private final String cnpj;
    private final char identificadorDeSocio;
    private final String nomeSocio;
    private final String cnpjCpf;
    private final String qualificacaoSocio;
    private final String percentualCapitalSocial;
    private final String dataEntradaSociedade;
    private final String codigoPais;
    private final String nomePaisSocio;
    private final String cpfRepresentanteLegal;
    private final String nomeRepresentante;
    private final String qualificacaoRepresentante;
    private final String filler;
    private final char fimDeRegistro;

    public static Socio parse(final byte[] source) {
        return new Socio(
            new String(copyOfRange(source, 0, 1)).charAt(0),
            new String(copyOfRange(source, 1, 2)).charAt(0),
            new String(copyOfRange(source, 2, 3)).charAt(0),
            new String(copyOfRange(source, 3, 17)).trim(),
            new String(copyOfRange(source, 17, 18)).charAt(0),
            new String(copyOfRange(source, 18, 168)).trim(),
            new String(copyOfRange(source, 168, 182)).trim(),
            new String(copyOfRange(source, 182, 184)).trim(),
            new String(copyOfRange(source, 184, 189)).trim(),
            new String(copyOfRange(source, 189, 197)).trim(),
            new String(copyOfRange(source, 197, 200)).trim(),
            new String(copyOfRange(source, 200, 270)).trim(),
            new String(copyOfRange(source, 270, 281)).trim(),
            new String(copyOfRange(source, 281, 341)).trim(),
            new String(copyOfRange(source, 341, 343)).trim(),
            new String(copyOfRange(source, 343, 1198)).trim(),
            new String(copyOfRange(source, 1198, 1199)).charAt(0)
        );
    }

    public Socio(char tipoDoRegistro, char indicadorFullDiario, char tipoAtualização, String cnpj, char identificadorDeSocio, String nomeSocio, String cnpjCpf, String qualificacaoSocio, String percentualCapitalSocial, String dataEntradaSociedade, String codigoPais, String nomePaisSocio, String cpfRepresentanteLegal, String nomeRepresentante, String qualificacaoRepresentante, String filler, char fimDeRegistro) {
        this.tipoDoRegistro = tipoDoRegistro;
        this.indicadorFullDiario = indicadorFullDiario;
        this.tipoAtualização = tipoAtualização;
        this.cnpj = cnpj;
        this.identificadorDeSocio = identificadorDeSocio;
        this.nomeSocio = nomeSocio;
        this.cnpjCpf = cnpjCpf;
        this.qualificacaoSocio = qualificacaoSocio;
        this.percentualCapitalSocial = percentualCapitalSocial;
        this.dataEntradaSociedade = dataEntradaSociedade;
        this.codigoPais = codigoPais;
        this.nomePaisSocio = nomePaisSocio;
        this.cpfRepresentanteLegal = cpfRepresentanteLegal;
        this.nomeRepresentante = nomeRepresentante;
        this.qualificacaoRepresentante = qualificacaoRepresentante;
        this.filler = filler;
        this.fimDeRegistro = fimDeRegistro;
    }

    public char getTipoDoRegistro() {
        return tipoDoRegistro;
    }

    public char getIndicadorFullDiario() {
        return indicadorFullDiario;
    }

    public char getTipoAtualização() {
        return tipoAtualização;
    }

    public String getCnpj() {
        return cnpj;
    }

    public char getIdentificadorDeSocio() {
        return identificadorDeSocio;
    }

    public String getNomeSocio() {
        return nomeSocio;
    }

    public String getCnpjCpf() {
        return cnpjCpf;
    }

    public String getQualificacaoSocio() {
        return qualificacaoSocio;
    }

    public String getPercentualCapitalSocial() {
        return percentualCapitalSocial;
    }

    public String getDataEntradaSociedade() {
        return dataEntradaSociedade;
    }

    public String getCodigoPais() {
        return codigoPais;
    }

    public String getNomePaisSocio() {
        return nomePaisSocio;
    }

    public String getCpfRepresentanteLegal() {
        return cpfRepresentanteLegal;
    }

    public String getNomeRepresentante() {
        return nomeRepresentante;
    }

    public String getQualificacaoRepresentante() {
        return qualificacaoRepresentante;
    }

    public String getFiller() {
        return filler;
    }

    public char getFimDeRegistro() {
        return fimDeRegistro;
    }
}

File: src/main/kotlin/model/Socio.kt
package model

import com.sun.xml.internal.ws.spi.db.BindingContextFactory.LOGGER
import java.util.Arrays.copyOfRange
import java.util.logging.Level

class Socio constructor() : InfoCnpj {
    var tipoDoRegistro: Char? = null
    var indicadorFullDiario: Char? = null
    var tipoAtualização: Char? = null
    lateinit var cnpj: String
    var identificadorDeSocio: Char? = null
    lateinit var nomeSocio: String
    lateinit var cnpjCpf: String
    lateinit var qualificacaoSocio: String
    lateinit var percentualCapitalSocial: String
    lateinit var dataEntradaSociedade: String
    lateinit var codigoPais: String
    lateinit var nomePaisSocio: String
    lateinit var cpfRepresentanteLegal: String
    lateinit var nomeRepresentante: String
    lateinit var qualificacaoRepresentante: String
    lateinit var filler: String
    var fimDeRegistro: Char? = null

    fun parse(source: ByteArray) : Socio {

        LOGGER.log(Level.INFO, String.format("Trying to parse Socio instance - %s ", String(copyOfRange(source, 18, 168))))

        return Socio(
                String(copyOfRange(source, 0, 1))[0],
                String(copyOfRange(source, 1, 2))[0],
                String(copyOfRange(source, 2, 3))[0],
                String(copyOfRange(source, 3, 17)),
                String(copyOfRange(source, 17, 18))[0],
                String(copyOfRange(source, 18, 168)),
                String(copyOfRange(source, 168, 182)),
                String(copyOfRange(source, 182, 184)),
                String(copyOfRange(source, 184, 189)),
                String(copyOfRange(source, 189, 197)),
                String(copyOfRange(source, 197, 200)),
                String(copyOfRange(source, 200, 270)),
                String(copyOfRange(source, 270, 281)),
                String(copyOfRange(source, 281, 341)),
                String(copyOfRange(source, 341, 343)),
                String(copyOfRange(source, 343, 1198)),
                String(copyOfRange(source, 1198, 1199))[0]
        )
    }

    constructor(tipoDoRegistro: Char, indicadorFullDiario: Char,
              tipoAtualização: Char, cnpj: String,
              identificadorDeSocio: Char, nomeSocio: String,
              cnpjCpf: String, qualificacaoSocio: String,
              percentualCapitalSocial: String, dataEntradaSociedade: String,
              codigoPais: String, nomePaisSocio: String,
              cpfRepresentanteLegal: String, nomeRepresentante: String,
              qualificacaoRepresentante: String, filler: String,
              fimDeRegistro: Char): this() {
        this.tipoDoRegistro = tipoDoRegistro
        this.indicadorFullDiario = indicadorFullDiario
        this.tipoAtualização = tipoAtualização
        this.cnpj = cnpj
        this.identificadorDeSocio = identificadorDeSocio
        this.nomeSocio = nomeSocio
        this.cnpjCpf = cnpjCpf
        this.qualificacaoSocio = qualificacaoSocio
        this.percentualCapitalSocial = percentualCapitalSocial
        this.dataEntradaSociedade = dataEntradaSociedade
        this.codigoPais = codigoPais
        this.nomePaisSocio = nomePaisSocio
        this.cpfRepresentanteLegal = cpfRepresentanteLegal
        this.nomeRepresentante = nomeRepresentante
        this.qualificacaoRepresentante = qualificacaoRepresentante
        this.filler = filler
        this.fimDeRegistro = fimDeRegistro
    }


}